File: app/src/main/java/com/mgaetan89/showsrage/network/SickRageApi.java
package com.mgaetan89.showsrage.network;

import android.content.SharedPreferences;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.mgaetan89.showsrage.Constants;
import com.mgaetan89.showsrage.model.Indexer;
import com.squareup.okhttp.Authenticator;
import com.squareup.okhttp.Credentials;
import com.squareup.okhttp.OkHttpClient;
import com.squareup.okhttp.Request;
import com.squareup.okhttp.Response;

import java.io.IOException;
import java.net.Proxy;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;

import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSession;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

import retrofit.RequestInterceptor;
import retrofit.RestAdapter;
import retrofit.client.OkClient;

public final class SickRageApi implements RequestInterceptor {
	private String apiKey = "";

	@NonNull
	private String apiUrl = "";

	@Nullable
	/* package */ String credentials = null;

	@NonNull
	private static final SickRageApi INSTANCE = new SickRageApi();

	private OkHttpClient okHttpClient = null;

	@NonNull
	private String path = "";

	private SickRageServices services = null;

	@NonNull
	private String webRoot = "";

	private SickRageApi() {
	}

	@NonNull
	public static SickRageApi getInstance() {
		return INSTANCE;
	}

	@NonNull
	public String getApiUrl() {
		return buildFullApiUrl(this.apiUrl, this.path, this.apiKey);
	}

	@NonNull
	public String getBannerUrl(int indexerId, @Nullable Indexer indexer) {
		if (indexer == null) {
			return this.getApiUrl() + "?cmd=show.getbanner";
		}

		return String.format("%s?cmd=show.getbanner&%s=%d", this.getApiUrl(), indexer.getParamName(), indexerId);
	}

	@NonNull
	public String getFanArtUrl(int indexerId, @Nullable Indexer indexer) {
		if (indexer == null) {
			return this.getApiUrl() + "?cmd=show.getfanart";
		}

		return String.format("%s?cmd=show.getfanart&%s=%d", this.getApiUrl(), indexer.getParamName(), indexerId);
	}

	@NonNull
	public OkHttpClient getOkHttpClient(boolean useSelfSignedCertificate) {
		if (this.okHttpClient == null) {
			this.okHttpClient = new OkHttpClient();

			this.setAuthenticator();
		}

		this.trustAllCertificates(useSelfSignedCertificate);

		return this.okHttpClient;
	}

	@NonNull
	public String getPosterUrl(int indexerId, @Nullable Indexer indexer) {
		if (indexer == null) {
			return this.getApiUrl() + "?cmd=show.getposter";
		}

		return String.format("%s?cmd=show.getposter&%s=%d", this.getApiUrl(), indexer.getParamName(), indexerId);
	}

	public SickRageServices getServices() {
		return this.services;
	}

	@NonNull
	public String getVideosUrl() {
		return this.apiUrl + "videos";
	}

	public void init(@NonNull SharedPreferences preferences) {
		boolean useHttps = preferences.getBoolean("use_https", false);
		String address = preferences.getString("server_address", "");
		String portNumber = preferences.getString("server_port_number", "");
		boolean selfSignedCertificate = preferences.getBoolean("self_signed_certificate", false);

		this.apiKey = preferences.getString("api_key", "");
		this.apiUrl = buildApiUrl(useHttps, address, portNumber);
		this.credentials = getCredentials(
				preferences.getBoolean("basic_auth", false),
				preferences.getString("server_username", null),
				preferences.getString("server_password", null)
		);
		this.path = getApiPath(preferences.getString("server_path", ""));
		this.webRoot = getWebRoot(this.path);

		RestAdapter.Builder builder = new RestAdapter.Builder();
		builder.setClient(new OkClient(this.getOkHttpClient(selfSignedCertificate)));
		builder.setEndpoint(this.apiUrl);
		builder.setRequestInterceptor(this);
		builder.setLogLevel(Constants.NETWORK_LOG_LEVEL);

		this.services = builder.build().create(SickRageServices.class);
	}

	@Override
	public void intercept(RequestFacade request) {
		request.addEncodedPathParam("api_path", this.path);
		request.addPathParam("api_key", this.apiKey);
		request.addEncodedPathParam("web_root", this.webRoot);
	}

	@NonNull
	private static String buildApiUrl(boolean useHttps, String address, String portNumber) {
		// Retrofit requires a non-empty endpoint
		// So we use the local url
		if (address.isEmpty()) {
			return "http://127.0.0.1/";
		}

		StringBuilder builder = new StringBuilder();

		if (useHttps) {
			builder.append("https://");
		} else {
			builder.append("http://");
		}

		builder.append(address);

		if (!portNumber.isEmpty()) {
			builder.append(":").append(portNumber);
		}

		builder.append("/");

		return builder.toString();
	}

	@NonNull
	private static String buildFullApiUrl(String apiUrl, String path, String apiKey) {
		StringBuilder builder = new StringBuilder();
		builder.append(apiUrl);

		if (!path.isEmpty()) {
			builder.append(path).append("/");
		}

		if (!apiKey.isEmpty()) {
			builder.append(apiKey).append("/");
		}

		return builder.toString();
	}

	@NonNull
	/* package */ static String getApiPath(String apiPath) {
		if (apiPath == null || apiPath.isEmpty()) {
			return "";
		}

		return apiPath.replaceAll("^/+|/$+", "");
	}

	@Nullable
	/* package */ static String getCredentials(boolean useBasicAuthentication, String username, String password) {
		if (!useBasicAuthentication) {
			return null;
		}

		if (username == null || username.isEmpty()) {
			return null;
		}

		if (password == null || password.isEmpty()) {
			return null;
		}

		return Credentials.basic(username, password);
	}

	@NonNull
	/* package */ static String getWebRoot(String apiPath) {
		if (apiPath == null || apiPath.isEmpty()) {
			return "";
		}

		String path = apiPath;

		// Add a leading /
		if (!path.startsWith("/")) {
			path = "/" + path;
		}

		// Add a trailing /
		if (!path.endsWith("/")) {
			path += "/";
		}

		// If the path ends with /api/, we ignore that last segment
		if (path.endsWith("/api/")) {
			return path.substring(1, path.length() - 4);
		}

		return apiPath.replaceAll("^/+|/$+", "") + "/";
	}

	private void setAuthenticator() {
		this.okHttpClient.setAuthenticator(new Authenticator() {
			@Override
			public Request authenticate(Proxy proxy, Response response) throws IOException {
				String credentials = SickRageApi.this.credentials;

				if (credentials != null && !credentials.isEmpty()) {
					return response.request().newBuilder()
							.header("Authorization", credentials)
							.build();
				}

				return null;
			}

			@Override
			public Request authenticateProxy(Proxy proxy, Response response) throws IOException {
				return null;
			}
		});
	}

	private void trustAllCertificates(boolean useSelfSignedCertificate) {
		if (!useSelfSignedCertificate) {
			return;
		}

		TrustManager[] trustAllCerts = new TrustManager[]{
				new X509TrustManager() {
					@Override
					public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
					}

					@Override
					public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
					}

					@Override
					public X509Certificate[] getAcceptedIssuers() {
						return null;
					}
				}
		};

		SSLContext sslContext = null;
		try {
			sslContext = SSLContext.getInstance("SSL");
			sslContext.init(null, trustAllCerts, new SecureRandom());
		} catch (KeyManagementException | NoSuchAlgorithmException exception) {
			exception.printStackTrace();
		}

		if (sslContext != null) {
			this.okHttpClient.setSslSocketFactory(sslContext.getSocketFactory());
		}

		this.okHttpClient.setHostnameVerifier(new HostnameVerifier() {
			@Override
			public boolean verify(String hostname, SSLSession session) {
				return true;
			}
		});
	}
}

File: app/src/main/kotlin/com/mgaetan89/showsrage/network/SickRageApi.kt
package com.mgaetan89.showsrage.network

import android.content.SharedPreferences
import com.google.gson.ExclusionStrategy
import com.google.gson.FieldAttributes
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.mgaetan89.showsrage.BuildConfig
import com.mgaetan89.showsrage.extension.getApiKey
import com.mgaetan89.showsrage.extension.getPortNumber
import com.mgaetan89.showsrage.extension.getServerAddress
import com.mgaetan89.showsrage.extension.getServerPassword
import com.mgaetan89.showsrage.extension.getServerPath
import com.mgaetan89.showsrage.extension.getServerUsername
import com.mgaetan89.showsrage.extension.useBasicAuth
import com.mgaetan89.showsrage.extension.useHttps
import com.mgaetan89.showsrage.extension.useSelfSignedCertificate
import com.mgaetan89.showsrage.model.Indexer
import com.squareup.okhttp.Authenticator
import com.squareup.okhttp.Credentials
import com.squareup.okhttp.OkHttpClient
import com.squareup.okhttp.Request
import com.squareup.okhttp.Response
import io.realm.RealmObject
import retrofit.RequestInterceptor
import retrofit.RestAdapter
import retrofit.client.OkClient
import retrofit.converter.GsonConverter
import java.io.IOException
import java.net.Proxy
import java.security.KeyManagementException
import java.security.NoSuchAlgorithmException
import java.security.SecureRandom
import java.security.cert.CertificateException
import java.security.cert.X509Certificate
import java.util.concurrent.TimeUnit
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager

class SickRageApi private constructor() : RequestInterceptor {
	private var apiKey = ""
	private var apiUrl = ""
	private var credentials: String? = null
	private var okHttpClient: OkHttpClient? = null
	private var path = ""
	var services: SickRageServices? = null
		private set
	val videosUrl: String
		get() = this.apiUrl + "videos"
	private var webRoot = ""

	fun getApiUrl() = buildFullApiUrl(this.apiUrl, this.path, this.apiKey)

	fun getBannerUrl(indexerId: Int, indexer: Indexer?): String {
		if (indexer == null) {
			return "%s?cmd=show.getbanner".format(this.getApiUrl())
		}

		return "%s?cmd=show.getbanner&%s=%d".format(this.getApiUrl(), indexer.paramName, indexerId)
	}

	fun getFanArtUrl(indexerId: Int, indexer: Indexer?): String {
		if (indexer == null) {
			return "%s?cmd=show.getfanart".format(this.getApiUrl())
		}

		return "%s?cmd=show.getfanart&%s=%d".format(this.getApiUrl(), indexer.paramName, indexerId)
	}

	fun getOkHttpClient(useSelfSignedCertificate: Boolean): OkHttpClient {
		val okHttpClient = this.okHttpClient ?: OkHttpClient().apply {
			this.setConnectTimeout(30, TimeUnit.SECONDS)
			this.setReadTimeout(30, TimeUnit.SECONDS)
		}

		this.setAuthenticator()
		this.trustAllCertificates(useSelfSignedCertificate)

		this.okHttpClient = okHttpClient

		return okHttpClient
	}

	fun getPosterUrl(indexerId: Int, indexer: Indexer?): String {
		if (indexer == null) {
			return "%s?cmd=show.getposter".format(this.getApiUrl())
		}

		return "%s?cmd=show.getposter&%s=%d".format(this.getApiUrl(), indexer.paramName, indexerId)
	}

	fun init(preferences: SharedPreferences) {
		val useHttps = preferences.useHttps()
		val address = preferences.getServerAddress()
		val portNumber = preferences.getPortNumber()
		val selfSignedCertificate = preferences.useSelfSignedCertificate()
		val logLevel = if (BuildConfig.DEBUG) RestAdapter.LogLevel.FULL else RestAdapter.LogLevel.NONE

		this.apiKey = preferences.getApiKey()
		this.apiUrl = buildApiUrl(useHttps, address, portNumber)
		this.credentials = getCredentials(
				preferences.useBasicAuth(),
				preferences.getServerUsername(),
				preferences.getServerPassword()
		)
		this.path = preferences.getServerPath()
		this.webRoot = getWebRoot(this.path)

		val builder = RestAdapter.Builder()
		builder.setClient(OkClient(this.getOkHttpClient(selfSignedCertificate)))
		builder.setConverter(gsonConverter)
		builder.setEndpoint(this.apiUrl)
		builder.setRequestInterceptor(this)
		builder.setLogLevel(logLevel)

		this.services = builder.build().create(SickRageServices::class.java)
	}

	override fun intercept(request: RequestInterceptor.RequestFacade) {
		request.addEncodedPathParam("api_path", this.path)
		request.addPathParam("api_key", this.apiKey)
		request.addEncodedPathParam("web_root", this.webRoot)
	}

	private fun setAuthenticator() {
		this.okHttpClient?.authenticator = object : Authenticator {
			@Throws(IOException::class)
			override fun authenticate(proxy: Proxy, response: Response): Request? {
				val credentials = this@SickRageApi.credentials

				if (!credentials.isNullOrEmpty()) {
					return response.request().newBuilder().header("Authorization", credentials).build()
				}

				return null
			}

			@Throws(IOException::class)
			override fun authenticateProxy(proxy: Proxy, response: Response) = null
		}
	}

	private fun trustAllCertificates(useSelfSignedCertificate: Boolean) {
		if (!useSelfSignedCertificate) {
			return
		}

		val trustAllCerts = arrayOf<TrustManager>(object : X509TrustManager {
			@Throws(CertificateException::class)
			override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) = Unit

			@Throws(CertificateException::class)
			override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) = Unit

			override fun getAcceptedIssuers() = null
		})

		var sslContext: SSLContext? = null
		try {
			sslContext = SSLContext.getInstance("SSL").apply {
				this.init(null, trustAllCerts, SecureRandom())
			}
		} catch (exception: KeyManagementException) {
			exception.printStackTrace()
		} catch (exception: NoSuchAlgorithmException) {
			exception.printStackTrace()
		}

		sslContext?.let {
			this.okHttpClient?.sslSocketFactory = it.socketFactory
		}

		this.okHttpClient?.setHostnameVerifier { _, _ -> true }
	}

	companion object {
		val gson: Gson by lazy {
			GsonBuilder()
					.setExclusionStrategies(object : ExclusionStrategy {
						override fun shouldSkipField(f: FieldAttributes) = f.declaringClass == RealmObject::class.java

						override fun shouldSkipClass(clazz: Class<*>) = false
					})
					.create()
		}
		private val gsonConverter by lazy { GsonConverter(gson) }
		val instance by lazy { SickRageApi() }

		private fun buildApiUrl(useHttps: Boolean, address: String, portNumber: String): String {
			// Retrofit requires a non-empty endpoint
			// So we use the local url
			if (address.isEmpty()) {
				return "http://127.0.0.1/"
			}

			val builder = StringBuilder()
			builder.append(if (useHttps) "https://" else "http://")
			builder.append(address)

			if (portNumber.isNotEmpty()) {
				builder.append(":").append(portNumber)
			}

			builder.append("/")

			return builder.toString()
		}

		private fun buildFullApiUrl(apiUrl: String, path: String, apiKey: String): String {
			val builder = StringBuilder()
			builder.append(apiUrl)

			if (path.isNotEmpty()) {
				builder.append(path).append("/")
			}

			if (apiKey.isNotEmpty()) {
				builder.append(apiKey).append("/")
			}

			return builder.toString()
		}

		fun getCredentials(useBasicAuthentication: Boolean, username: String?, password: String?): String? {
			if (!useBasicAuthentication || username.isNullOrEmpty() || password.isNullOrEmpty()) {
				return null
			}

			return Credentials.basic(username, password)
		}

		fun getWebRoot(apiPath: String?): String {
			if (apiPath.isNullOrEmpty()) {
				return ""
			}

			var path = apiPath!!

			// Add a leading /
			if (!path.startsWith("/")) {
				path = "/%s".format(path)
			}

			// Add a trailing /
			if (!path.endsWith("/")) {
				path += "/"
			}

			// If the path ends with /api/, we ignore that last segment
			if (path.endsWith("/api/")) {
				return path.substring(1, path.length - 4)
			}

			return apiPath.replace("^/+|/$+".toRegex(), "") + "/"
		}

		fun setCredentials(credentials: String?) {
			SickRageApi.instance.credentials = credentials
		}
	}
}
